src/mixins/createI18nMixin.ts

```diff
@@ -0,0 +1,150 @@
+	 * is updated.
+	 *
+	 * @param bundle
+	 * The required bundle object for which available locale messages should be loaded.
+	 *
+	 * @return
+	 * The localized messages, along with a `format` method for formatting ICU-formatted templates.
+	 */
+	localizeBundle(bundle: Bundle<M>): LocalizedMessages<M>;
+}
+
+export interface I18nFactory extends ComposeFactory<I18nMixin<Messages>, WidgetOptions<WidgetState, I18nProperties>> {}
+
+export interface I18nProperties extends WidgetProperties {
+	/**
+	 * The locale for the widget. Is not specified, then the root locale (as determined by `dojo-i18n`) is assumed.
+	 * If specified, the widget's node will have a `data-locale` property set to the locale, in order to facilitate
+	 * styling localized components if the use case arises.
+	 */
+	locale?: string;
+
+	/**
+	 * An optional flag indicating the widget's text direction. If `true`, then the underlying node's `dir`
+	 * property is set to "rtl". If it is `false`, then the `dir` property is set to "ltr". Otherwise, the property
+	 * is not set.
+	 */
+	rtl?: boolean;
+}
+
+/**
+ * @private
+ * An internal helper interface for defining locale and text direction attributes on widget nodes.
+ */
+interface I18nVNodeProperties extends VNodeProperties {
+	'data-locale': string | null;
+	dir: string | null;
+}
+
+export type I18nWidget<M extends Messages, P extends I18nProperties> = I18nMixin<M> & Widget<WidgetState, I18nProperties>;
+
+export type LocalizedMessages<T extends Messages> = T & {
+	/**
+	 * Formats an ICU-formatted message template for the represented bundle.
+	 *
+	 * @param key
+	 * The message key.
+	 *
+	 * @param options
+	 * The values to pass to the formatter.
+	 *
+	 * @return
+	 * The formatted string.
+	 */
+	format(key: string, options?: any): string;
+}
+
+/**
+ * @private
+ * Return the cached dictionary for the specified bundle and locale, if it exists. If the requested dictionary does not
+ * exist, then load it and update the instance's state with the appropriate messages.
+ *
+ * @param instance
+ * The widget instance.
+ *
+ * @param bundle
+ * The bundle for which to load a locale-specific dictionary.
+ *
+ * @return
+ * The locale-specific dictionary, if it has already been loaded and cached.
+ */
+function getLocaleMessages(instance: I18nWidget<Messages, I18nProperties>, bundle: Bundle<Messages>): Messages | void {
+	const { properties } = instance;
+	const locale = properties['locale'] || i18n.locale;
+	const localeMessages = getCachedMessages(bundle, locale);
+
+	if (localeMessages) {
+		return localeMessages;
+	}
+
+	i18n(bundle, locale).then(() => {
+		instance.invalidate();
+	});
+}
+
+const createI18nMixin: I18nFactory = compose<I18nMixin<Messages>, WidgetOptions<WidgetState, I18nProperties>>({
+	nodeAttributes: [
+		function (this: I18nWidget<Messages, I18nProperties>, attributes: VNodeProperties): VNodeProperties {
+			const properties = {
```

Comment:
lots of `properties` going on in here, can we call the variable for `VNodeProperties` `vNodeProperties` or similar?